internal/app/db.go

```diff
@@ -0,0 +1,331 @@
+package app
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/schollz/progressbar/v3"
+	"github.com/tablelandnetwork/basin-cli/pkg/pgrepl"
+	"golang.org/x/exp/slog"
+)
+
+// Column represents a column in a table being replicated.
+type Column struct {
+	Name, Typ         string
+	IsNull, IsPrimary bool
+}
+
+// DBManager manages a duckdb database.
+type DBManager struct {
+	db        *sql.DB
+	dbDir     string
+	dbFname   string
+	table     string
+	createdAT time.Time
+	cols      []Column
+	winSize   time.Duration
+	uploader  *BasinUploader
+}
+
+// NewDBManager creates a new DBManager.
+func NewDBManager(dbDir, table string, cols []Column, winSize time.Duration, uploader *BasinUploader) *DBManager {
+	return &DBManager{
+		dbDir:    dbDir,
+		table:    table,
+		cols:     cols,
+		winSize:  winSize,
+		uploader: uploader,
+	}
+}
+
+// Close closes the current db.
+func (dbm *DBManager) Close() error {
+	return dbm.db.Close()
+}
+
+// queryFromWAL creates a query for a WAL TX records.
+// (todo): error handling and remapping of types.
+func (dbm *DBManager) queryFromWAL(tx *pgrepl.Tx) string {
+	var queries []string
+	for _, r := range tx.Records {
+		columnNames := []string{}
+		vals := []string{}
+		for _, c := range r.Columns {
+			columnNames = append(columnNames, c.Name)
+			vals = append(vals, strings.ReplaceAll(string(c.Value), "\"", ""))
+		}
+
+		cols := strings.Join(columnNames, ", ")
+		valsStr := strings.Join(vals, ", ")
+		query := fmt.Sprintf(
```

Comment:
you don't have to create multiple insert stmts (one for each record), you can do all into one statement, e.g. `INSERT INTO %s (%s) values (%s), (%s), (%s), ...`

Reply:
nice! i didn't know that you could write an insert statement like that